src/VisualStudio/Core/Def/Implementation/ProjectSystem/Legacy/AbstractLegacyProject.cs

```diff
@@ -1,135 +1,189 @@
 
 namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.Legacy
 {
     /// <summary>
     /// Base type for legacy C# and VB project system shim implementations.
-    /// These legachy shims are based on legacy project system interfaces defined in csproj/msvbprj.
+    /// These legacy shims are based on legacy project system interfaces defined in csproj/msvbprj.
     /// </summary>
-    internal abstract partial class AbstractLegacyProject : AbstractProject
+    internal abstract partial class AbstractLegacyProject : ForegroundThreadAffinitizedObject
     {
+        public IVsHierarchy Hierarchy { get; }
+        protected VisualStudioProject VisualStudioProject { get; }
+        internal VisualStudioProjectOptionsProcessor VisualStudioProjectOptionsProcessor { get; set; }
+        protected IProjectCodeModel ProjectCodeModel { get; set; }
+        protected VisualStudioWorkspace Workspace { get; }
+
+        #region Mutable fields that should only be used from the UI thread
+
+        private readonly VsENCRebuildableProjectImpl _editAndContinueProject;
+
+        private readonly SolutionEventsBatchScopeCreator _batchScopeCreator;
+
+        #endregion
+
         public AbstractLegacyProject(
-            VisualStudioProjectTracker projectTracker,
-            Func<ProjectId, IVsReportExternalErrors> reportExternalErrorCreatorOpt,
             string projectSystemName,
             IVsHierarchy hierarchy,
             string language,
             IServiceProvider serviceProvider,
-            VisualStudioWorkspaceImpl visualStudioWorkspaceOpt,
+            IThreadingContext threadingContext,
+            string externalErrorReportingPrefix,
             HostDiagnosticUpdateSource hostDiagnosticUpdateSourceOpt,
-            ICommandLineParserService commandLineParserServiceOpt = null)
-            : base(projectTracker,
-                  reportExternalErrorCreatorOpt,
-                  projectSystemName,
-                  projectFilePath: GetProjectFilePath(hierarchy),
-                  hierarchy: hierarchy,
-                  projectGuid: GetProjectIDGuid(hierarchy),
-                  language: language,
-                  serviceProvider: serviceProvider,
-                  visualStudioWorkspaceOpt: visualStudioWorkspaceOpt,
-                  hostDiagnosticUpdateSourceOpt: hostDiagnosticUpdateSourceOpt,
-                  commandLineParserServiceOpt: commandLineParserServiceOpt)
+            ICommandLineParserService commandLineParserServiceOpt)
+            : base(threadingContext)
         {
-            if (Hierarchy != null)
+            Contract.ThrowIfNull(hierarchy);
+
+            var componentModel = (IComponentModel)serviceProvider.GetService(typeof(SComponentModel));
+            Workspace = componentModel.GetService<VisualStudioWorkspace>();
+
+            var projectFilePath = hierarchy.GetProjectFilePath();
+
+            if (projectFilePath != null && !File.Exists(projectFilePath))
```

Comment:
>if (projectFilePath != null && !File.Exists(projectFilePath)) [](start = 12, length = 61)

Assuming this is a workaround for something that will go away one day?

Reply:
It's the workaround we've had for 7 years, apparently. (It just wasn't very obvious that the workaround indeed existed.) Hopefully AbstractLegacyProject goes away later in the Dev16 wave.